# Projeto 1 - Ciência dos Dados

Nome: Augusto Ferreira

Nome: Diogo Diniz

Nome: João Andrade

Atenção: Serão permitidos grupos de três pessoas, mas com uma rubrica mais exigente. Grupos deste tamanho precisarão fazer um questionário de avaliação de trabalho em equipe

___
Carregando algumas bibliotecas:

In [50]:
%matplotlib inline
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import os

In [51]:
print('Esperamos trabalhar no diretório')
print(os.getcwd())

Esperamos trabalhar no diretório
c:\zTodoresto\Insper\Projeto1Cdados\cdados-p1


Carregando a base de dados com os tweets classificados manualmente:

In [52]:
train = pd.read_excel('dados_treino.xlsx')
train.head(5)

Mensagem   Unnamed: 1
0  Eu esperava um livro mais profundo pelas altas...     Negativo
1  Gostei muito do livro. Prende a nossa atenção ...     Positivo
2  Intediante, chato, o livro fica por dizer cois...     Negativo
3  Trilogia perfeita!!!!!!!! Recomendo!!!! Melhor...     Positivo
4  Comprei para presentear, então não posso avali...  Irrelevante

In [53]:
test = pd.read_excel('dados_teste.xlsx')
test.head(5)

Mensagem   Unnamed: 1
0  Solicitei meu pagamento em 4 x e quando fui ve...  irrelevante
1  Tradução péssima!!! Os editores e o tradutor d...  irrelevante
2  Peço que verifiquem e reparem este erro, acred...  irrelevante
3  Queria comprar a coleção... Aí fui antes numa ...  irrelevante
4  Comprei os boxes pela primeira vez e o produto...  irrelevante

___
## Classificador automático


Faça aqui uma descrição do seu assunto e o que considerou como relevante ou não relevante na classificação dos tweets (Target).

Foram considerados relevantes as informações que traziam comentários positivos ou negativos sobre o livro em si, e foram desconsiderados aqueles que falavam sobre problemas com a entrega, com o modelo digital, sobre preços e pagamentos.

___
### Montando um Classificador Naive-Bayes

Considerando apenas as mensagens da planilha Treinamento, ensine  seu classificador.

In [59]:
def limpa_texto(texto):
    caractere_especial = ['', '!', '@', '¨', '#', '$',  '%', '&', '*', '?', '"',',','.']
    texto_limpo = ''
    for letra in texto:
        if letra not in  caractere_especial:
            letra_nova = letra.lower()
            texto_limpo += letra_nova
    return texto_limpo

test.columns = ['comentario','tag']
test['comentario'] = test['comentario'].apply(limpa_texto)
test['tag'] = test['tag'].apply(limpa_texto)
test

comentario          tag
0    solicitei meu pagamento em 4 x e quando fui ve...  irrelevante
1    tradução péssima os editores e o tradutor devi...  irrelevante
2    peço que verifiquem e reparem este erro acredi...  irrelevante
3    queria comprar a coleção aí fui antes numa liv...  irrelevante
4    comprei os boxes pela primeira vez e o produto...  irrelevante
..                                                 ...          ...
245  olavo é um fanático religioso que encontrou pú...     negativo
246  os mapas estão em péssima qualidade e os capít...     negativo
247  sem dúvida esse é o pior livro que já li não p...     negativo
248  livro meio fantasioso  quer dizer que quase nã...     negativo
249  o livro exibe umas poucas técnicas de maneira ...     negativo

[250 rows x 2 columns]

In [75]:
test['tag'].value_counts()
k=0


def separador_palavras(dt):
    k=0
    texto=''

    while k < len(dt['comentario']):
        texto += dt['comentario'][k]

        k+=1

    return texto.split()

palavras = separador_palavras(test)
lista_palavras_total= pd.Series(palavras)
lista_palavras_total = lista_palavras_total.value_counts(True)
lista_palavras_total['que']


0.03424486742581664

___
### Verificando a performance do Classificador

Agora você deve testar o seu classificador com a base de Testes.

___
### Concluindo

___
### Qualidade do Classificador a partir de novas separações dos tweets entre Treinamento e Teste

Caso for fazer esse item do Projeto

___
## Aperfeiçoamento:

Trabalhos que conseguirem pelo menos conceito B vão evoluir em conceito dependendo da quantidade de itens avançados:

* IMPLEMENTOU outras limpezas e transformações que não afetem a qualidade da informação contida nos tweets. Ex: stemming, lemmatization, stopwords
* CONSIDEROU mais de duas categorias na variável Target e INCREMENTOU a quantidade de notícias, mantendo pelo menos 250 notícias por categoria (OBRIGATÓRIO PARA TRIOS, sem contar como item avançado)
* Para Target com duas categorias: CRIOU pelo menos quatro categorias intermediárias de relevância baseadas na probabilidade: ex.: muito relevante, relevante, neutro, irrelevante, muito irrelevante
* EXPLICOU porquê não pode usar o próprio classificador para gerar mais amostras de treinamento
* PROPÔS diferentes cenários para Naïve Bayes fora do contexto do projeto (pelo menos dois cenários, exceto aqueles já apresentados em sala pelos professores: por exemplo, filtro de spam)
* SUGERIU e EXPLICOU melhorias reais com indicações concretas de como implementar (indicar como fazer e indicar material de pesquisa)
* FEZ o item Qualidade do Classificador a partir de novas separações das Notícias entre Treinamento e Teste descrito no enunciado do projeto (OBRIGATÓRIO para conceitos A ou A+)

___
## Referências

[Naive Bayes and Text Classification](https://arxiv.org/pdf/1410.5329.pdf)  **Mais completo**

[A practical explanation of a Naive Bayes Classifier](https://monkeylearn.com/blog/practical-explanation-naive-bayes-classifier/) **Mais simples**